pkg/action/install.go

```diff
@@ -83,29 +83,98 @@ func CheckDependencies(ch *helmChart.Chart, reqs []*helmChart.Dependency) error
+// satisfied, they will be silently skipped.
+func (i *Install) Run(strategy solver.SolverStrategy, wantedChrt *helmChart.Chart, vals map[string]interface{},
+	settings *cli.EnvSettings, logger log.Logger) ([]*release.Release, error) {
+
+	// TODO obtain lock
+	// defer release lock
+
+	// create pkg with chart to be installed:
+	version := i.ChartPathOptions.Version
+	pinnedVer := pkg.Unknown
+	if i.Version == "" {
+		// no pinned ver, take the chart as a filler for fp:
+		version = wantedChrt.Metadata.Version
+	} else {
+		pinnedVer = pkg.Present
+	}
+
+	wantedPkg := pkg.NewPkg(i.ReleaseName, wantedChrt.Metadata.Name, version, i.Namespace, pkg.Unknown, pkg.Present, pinnedVer, i.ChartPathOptions.RepoURL)
 
-	if lvl >= 10 {
-		return nil, errors.Errorf("ABORTING: Nested recursion #%d. we don't have a SAT solver yet, chances are you are in a cycle!", lvl)
+	// get all releases
+	rels, err := i.GetAllReleases()
+	if err != nil {
+		return nil, err
 	}
 
-	if !i.NoSharedDeps {
-		if err := i.InstallAllSharedDeps(chrt, settings, logger, lvl); err != nil {
-			return nil, err
+	// get all repo entries
+	rf, err := repo.LoadFile(settings.RepositoryConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	s := solver.New(strategy)
+
+	err = BuildWorld(s.PkgDB, rf.Repositories, rels, wantedPkg, wantedChrt, settings, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	s.PkgDB.DebugPrintDB(logger)
+
+	// Promote optional deps to normal deps, depending on the strategy selected:
+	// TODO use wantedPkg instead of wantedPkgInDB once wantedPkg from local chart gets depRel correctly built
+	wantedPkgInDB := solver.PkgDBInstance.GetPackageByFingerprint(wantedPkg.GetFingerPrint())
+	switch i.OptionalDeps {
+	case OptionalDepsAll:
+		logger.Debugf("Promoting all optional deps of package %s to normal deps\n", wantedPkgInDB.GetFingerPrint())
+		// promote all optional deps of wanted package to normal deps:
+		wantedPkgInDB.DependsRel = append(wantedPkgInDB.DependsRel, wantedPkgInDB.DependsOptionalRel...)
+	case OptionalDepsNone:
+		logger.Debugf("Disregarding all optional deps of package %s\n", wantedPkgInDB.GetFingerPrint())
+	case OptionalDepsAsk:
+		logger.Debugf("Asking for each optional deps of package %s if they should be promoted\n", wantedPkgInDB.GetFingerPrint())
+		for _, rel := range wantedPkgInDB.DependsOptionalRel {
+			reader := bufio.NewReader(os.Stdin)
+			question := eyecandy.ESPrintf(settings.NoEmojis,
+				":red_question_mark:Install optional shared dependency \"%s\" of chart \"%s\"?",
+				rel.BaseFingerprint,
+				wantedPkgInDB.ChartName,
+			)
+			if promptBool(question, reader, logger) {
+				wantedPkgInDB.DependsRel = append(wantedPkgInDB.DependsRel, rel)
+			}
 		}
 	}
 
-	prefix := ""
-	if lvl > 0 {
-		prefix = fmt.Sprintf("%*s", lvl*2, "- ")
+	s.Solve(logger)
+
+	installedRels := make([]*release.Release, 0)
+	if s.IsSAT() {
+		for _, p := range s.PkgResultSet.ToInstall {
```

Comment:
I know I could figure this out by rereading and working through the code again but I think it's faster to ask as I'm going through this, what order will the packages come out of this in?

The order is important. For example, let's say there is a package A that has a dependency on controller B which has a dependency on CRD chart C (A -> B -> C). The install order needs to be C, B, A.

Reply:
Indeed, the order is important, and we aren't following it yet.

The solver returns a map of package fingerprints. Right now we just short them with `s.SortPkgSets()`, after separating them in packages to be installed, removed, and unchanged.

I suppose we should solve and then create the resulting graphs (as we have the dependency relations in the package structs). And then iterate through them, installing whatever it is needed. I don't think this is going to land in this PR, though.

(for CRDs, I would be glad to have pre-depends in the future).